pkg/certificate/signing_requestor.go

```diff
@@ -191,64 +221,106 @@ func (s *signingRequestorImpl) Issue(ctx context.Context, name string, ips []str
 		return errors.New("ips cannot be empty")
 	}
 
-	keyPEM, csrPEM, err := s.keyGenerator(ips)
-	if err != nil {
-		return errors.Wrapf(err, "error generating key and CSR data for %q and IPs %v", name, ips)
-	}
+	secretName := s.secretName(name)
 
 	newSecret := &corev1.Secret{
 		ObjectMeta: metav1.ObjectMeta{
-			Name:      s.secretName(name),
+			Name:      secretName,
 			Namespace: s.localNamespace,
 			Labels: map[string]string{
 				SigningRequestLabelKey: s.localClusterID,
 			},
 		},
-		Data: map[string][]byte{
-			PrivateKeyDataKey: keyPEM,
-			CSRDataKey:        csrPEM,
-		},
 	}
 
 	result, _, err := util.CreateOrUpdateWithOptions(ctx, util.CreateOrUpdateOptions[*unstructured.Unstructured]{
 		Client: resource.ForDynamic(s.localSecretClient),
 		Obj:    resource.MustToUnstructured(newSecret),
+		MutateOnCreate: func(obj *unstructured.Unstructured) (*unstructured.Unstructured, error) {
+			secret := resource.MustFromUnstructured(obj, &corev1.Secret{})
+
+			keyPEM, csrPEM, err := s.keyGenerator(ips)
+			if err != nil {
+				return nil, errors.Wrapf(err, "error generating key and CSR data for %q and IPs %v", name, ips)
+			}
+
+			secret.Data = map[string][]byte{
+				PrivateKeyDataKey: keyPEM,
+				CSRDataKey:        csrPEM,
+			}
+
+			return resource.MustToUnstructured(secret), nil
+		},
 		MutateOnUpdate: func(obj *unstructured.Unstructured) (*unstructured.Unstructured, error) {
 			existing := resource.MustFromUnstructured(obj, &corev1.Secret{})
-			if !bytes.Equal(existing.Data[CSRDataKey], newSecret.Data[CSRDataKey]) {
-				// The CSR data changed so clear the signed annotation, so it will be re-signed.
-				delete(existing.Annotations, RequestSignedLabelKey)
+
+			// Check if we need to regenerate the CSR by comparing the IPs
+			// We need to extract the IPs from the existing CSR to compare
+			needsNewCSR := false
+
+			if len(existing.Data[CSRDataKey]) == 0 {
+				// No existing CSR data, need to generate
+				needsNewCSR = true
+			} else {
+				// Parse existing CSR to extract IPs and compare
+				existingIPs, err := s.extractIPsFromCSR(existing.Data[CSRDataKey])
+				if err != nil {
+					// If we can't parse the existing CSR, generate a new one
+					needsNewCSR = true
+				} else {
+					// Compare IP lists
+					needsNewCSR = !s.ipListsEqual(existingIPs, ips)
+				}
 			}
 
-			existing.Data[PrivateKeyDataKey] = newSecret.Data[PrivateKeyDataKey]
-			existing.Data[CSRDataKey] = newSecret.Data[CSRDataKey]
+			if needsNewCSR {
+				keyPEM, csrPEM, err := s.keyGenerator(ips)
+				if err != nil {
+					return nil, errors.Wrapf(err, "error generating key and CSR data for %q and IPs %v", name, ips)
+				}
+
+				existing.Data[PrivateKeyDataKey] = keyPEM
+				existing.Data[CSRDataKey] = csrPEM
```

Comment:
As per suggested change above, we can keep this as:

```suggestion
			existing.Data[PrivateKeyDataKey] = newSecret.Data[PrivateKeyDataKey]
			existing.Data[CSRDataKey] = newSecret.Data[CSRDataKey]
```